File: Swift/SQLite.swift_example/Sources/repository.swift
import SQLite
import Foundation

enum RepositoryError: Error {
    case notFoundError(String)
    case castException
}

class Repository {
    let conn: Connection

    private let departments = Table("departments")
    private let departmentsId = Expression<Int>("id")
    private let departmentsName = Expression<String>("name")

    private let people = Table("people")
    private let peopleId = Expression<String>("id")
    private let peopleName = Expression<String>("name")
    private let peopleAddress = Expression<String>("address")
    private let peopleBirthDate = Expression<Date>("birth_date")

    private let employees = Table("employees")
    private let employeesId = Expression<String>("id")
    private let employeesDepartmentId = Expression<Int>("department_id")
    private let employeesPersonId = Expression<String>("person_id")
    private let employeesPosition = Expression<String>("position")

    init(connection: Connection) throws {
        conn = connection
        try initDb()
    }

    private func initDb() throws {
        _ = try conn.prepare("PRAGMA foreign_keys = ON;").scalar()

        try conn.run(departments.create(ifNotExists: true) {t in
            t.column(departmentsId, primaryKey: true)
            t.column(departmentsName)
        })

        try conn.run(people.create(ifNotExists: true) {t in
            t.column(peopleId, primaryKey: true)
            t.column(peopleName)
            t.column(peopleAddress)
            t.column(peopleBirthDate)
        })

        try conn.run(employees.create(ifNotExists: true) {t in
            t.column(employeesId, primaryKey: true)
            t.column(employeesDepartmentId)
            t.column(employeesPersonId)
            t.column(employeesPosition)
            t.foreignKey(employeesDepartmentId, references: departments, departmentsId, delete: .cascade)
            t.foreignKey(employeesPersonId, references: people, peopleId, delete: .cascade)
        })
    }

    func addDepartment(name: String) throws -> Department {
        let rowId = try conn.run(departments.insert(departmentsName <- name))
        return Department(id: Int(rowId), name: name)
    }

    func addPerson(name: String, address: String, birthDate: Date) throws -> Person {
        let id = NSUUID().uuidString
        try conn.run(people.insert(
            peopleId <- id,
            peopleName <- name,
            peopleAddress <- address,
            peopleBirthDate <- birthDate
        ))

        return Person(id: id, name: name, address: address, birthDate: birthDate)
    }

    func addEmployee(per: Person, dep: Department, position: Position) throws -> Employee {
        let id = NSUUID().uuidString
        try conn.run(employees.insert(
            employeesId <- id,
            employeesDepartmentId <- dep.id,
            employeesPersonId <- per.id,
            employeesPosition <- position.rawValue
        ))

        return Employee(id: id, department: dep, person: per, position: position)
    }

    func addEmployeeAndPerson(dep: Department, name: String, address: String, birthDate: Date, position: Position) throws -> Employee {
        let per = try addPerson(name: name, address: address, birthDate: birthDate)
        return try addEmployee(per: per, dep: dep, position: position)
    }

    func getDepartmentById(id: Int) throws -> Department? {
        if let row = try conn.pluck(departments.where(departmentsId == id)) {
            return Department(id: row[departmentsId], name: row[departmentsName])
        }

        return nil
    }

    func getPersonById(id: String) throws -> Person? {
        if let row = try conn.pluck(people.where(peopleId == id)) {
            return Person(id: row[peopleId], name: row[peopleName], address: row[peopleAddress], birthDate: row[peopleBirthDate])
        }

        return nil
    }

    func getEmployeeById(id: String) throws -> Employee? {
        if let row = try conn.pluck(employees.where(employeesId == id)) {
            if let per = try getPersonById(id: row[employeesPersonId]),
                let dep = try getDepartmentById(id: row[employeesDepartmentId]) {
                return Employee(id: row[employeesId], department: dep, person: per, position: Position(rawValue: row[employeesPosition]))
            }
        }

        return nil
    }

    func getAllDepartments() throws -> [Department] {
        var allDepartments: [Department] = []
        for row in try conn.prepare(departments) {
            allDepartments.append(Department(id: row[departmentsId], name: row[departmentsName]))
        }

        return allDepartments
    }

    func getAllPeople() throws -> [Person] {
        var allPeople: [Person] = []
        for row in try conn.prepare(people) {
            allPeople.append(Person(
                id: row[peopleId],
                name: row[peopleName],
                address: row[peopleAddress],
                birthDate: row[peopleBirthDate]
            ))
        }

        return allPeople
    }

    func getAllEmployees() throws -> [Employee] {
        var allEmployees: [Employee] = []
        for row in try conn.prepare(employees) {
            if let per = try getPersonById(id: row[employeesPersonId]),
                let dep = try getDepartmentById(id: row[employeesDepartmentId]) {
                allEmployees.append(Employee(id: row[employeesId], department: dep, person: per, position: Position(rawValue: row[employeesPosition])))
            } else {
                throw RepositoryError.notFoundError("one of employees foreign keys error")
            }
        }

        return allEmployees
    }

    func updateDepartment(dep: Department) throws {
        try conn.run(departments.where(departmentsId == dep.id).update(departmentsName <- dep.name))
    }

    func updatePerson(per: Person) throws {
        try conn.run(people.where(peopleId == per.id).update(
            peopleName <- per.name,
            peopleAddress <- per.address,
            peopleBirthDate <- per.birthDate
        ))
    }

    func updateEmployee(emp: Employee) throws {
        try conn.run(employees.where(employeesId == emp.id).update(
            employeesDepartmentId <- emp.department!.id,
            employeesPersonId <- emp.person!.id,
            employeesPosition <- emp.position!.rawValue
        ))
    }

    func delete<T>(entities: [T]) throws {
        if entities.count == 0 {
            return
        }

        if let collection = entities as? [Department] {
            try conn.run(departments.where(collection.map{d in d.id}.contains(departmentsId)).delete())
        } else if let collection = entities as? [Person] {
            try conn.run(people.where(collection.map{p in p.id}.contains(peopleId)).delete())
        } else if let collection = entities as? [Employee] {
            try conn.run(employees.where(collection.map{e in e.id}.contains(employeesId)).delete())
        } else {
            throw RepositoryError.castException
        }
    }
}

File: Go/Beego/repository/repository.go
package repository

import (
	"fmt"
	"sample/fields"
	"sample/models"
	"time"

	"github.com/beego/beego/v2/client/orm"
	"github.com/google/uuid"
)

var (
	_ Id = &models.Department{}
	_ Id = &models.Person{}
	_ Id = &models.Employee{}
)

type Repository struct {
	o orm.Ormer
}

func NewRepository(databaseAlias, databaseDriver, connectionString string) (*Repository, error) {
	orm.RegisterDataBase(databaseAlias, databaseDriver, connectionString)
	o := orm.NewOrm()
	if _, err := o.Raw("PRAGMA foreign_keys = ON;").Exec(); err != nil {
		return nil, err
	}

	return &Repository{o: o}, nil
}

func (r *Repository) AddDepartment(name string) (dep *models.Department, err error) {
	dep = &models.Department{Name: name}
	dep.Id, err = r.o.Insert(dep)
	return
}

func (r *Repository) AddPerson(name, address string, birthDate time.Time) (per *models.Person, err error) {
	per = &models.Person{Id: uuid.NewString(), Name: name, Address: address, BirthDate: birthDate}
	_, err = r.o.Insert(per)
	return
}

func (r *Repository) AddEmployee(person *models.Person, department *models.Department, position fields.PositionField) (emp *models.Employee, err error) {
	emp = &models.Employee{Id: uuid.NewString(), DepartmentId: department.Id, Department: department, PersonId: person.Id, Person: person, Position: position}
	_, err = r.o.Insert(emp)
	return
}

func (r *Repository) AddEmployeeWithPerson(
	department *models.Department,
	name, address string,
	birthDate time.Time,
	position fields.PositionField,
) (emp *models.Employee, err error) {
	var per *models.Person
	per, err = r.AddPerson(name, address, birthDate)
	if err != nil {
		return
	}

	return r.AddEmployee(per, department, position)
}

func (r *Repository) Update(obj any) (err error) {
	_, err = r.o.Update(obj)
	return
}

func (r *Repository) GetAllDepartments() (departments []*models.Department, err error) {
	_, err = r.o.QueryTable(&models.Department{}).All(&departments)
	return
}

func (r *Repository) GetAllPeople() (people []*models.Person, err error) {
	_, err = r.o.QueryTable(&models.Person{}).All(&people)
	return
}

func (r *Repository) GetAllEmployees() (employees []*models.Employee, err error) {
	if _, err = r.o.QueryTable(&models.Employee{}).All(&employees); err != nil {
		return
	}

	var (
		departments    []*models.Department
		departmentsMap = map[int64]*models.Department{}
		people         []*models.Person
		peopleMap      = map[string]*models.Person{}
	)

	if departments, err = r.GetAllDepartments(); err != nil {
		return
	}

	if people, err = r.GetAllPeople(); err != nil {
		return
	}

	for _, dep := range departments {
		departmentsMap[dep.Id] = dep
	}

	for _, per := range people {
		peopleMap[per.Id] = per
	}

	for _, emp := range employees {
		emp.Department = departmentsMap[emp.DepartmentId]
		emp.Person = peopleMap[emp.PersonId]
	}

	return
}

type Id interface {
	IdFieldName() string
	IdValue() any
}

func (r *Repository) Delete(obj Id, table any) error {
	_, err := r.o.QueryTable(table).Filter(obj.IdFieldName(), obj.IdValue()).Delete()
	return err
}

func (r *Repository) DeleteMany(objects []Id, table any) error {
	if len(objects) == 0 {
		return nil
	}

	ids := []any{}
	for _, obj := range objects {
		ids = append(ids, obj.IdValue())
	}

	_, err := r.o.QueryTable(table).Filter(fmt.Sprintf("%s__in", objects[0].IdFieldName()), ids...).Delete()
	return err
}
